### PR TITLE
[CDAP-20648] Create CredentialProfile and CredentialIdentity stores

### PR DESCRIPTION
**Pull request by dli357**
Encryption SPI will come in a follow-up PR for sensitive data stored in the StructuredTables.

### CODE REVIEW

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/SecureMultiStore.java
Comment (rmstar) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/SecureMultiStore.java]: nit: this can just be named get()

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/Util.java
Comment (rmstar) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/Util.java]: Better to store the compiled regex (Pattern) here and match against that

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/CredentialIdentityStore.java
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/CredentialIdentityStore.java]: in general, injected constructors should be package private or private to ensure somebody doesn't `new` it except in unit tests.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/SecureMultiStore.java
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/SecureMultiStore.java]: one of the problems with having an abstraction over writing to two different systems is that the operation isn't actually atomic, which can cause all sorts of issues. For example, the secure store write can complete while the DB write fails. The overall create fails, but now there is dirty leftover data so it's more like a half-write. It can't be fully fixed by deleting the secure data on failure, as the delete can fail as well. It's also open to race conditions, if two threads try to create the same resource at the same time.

I would recommend separating out the operations instead of trying to put them together. In other words, the caller needs to create the resource first, and can only update metadata for an existing resource.

Comment (albertshau) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/SecureMultiStore.java]: alternatively if it's possible to just use the secure store instead of secure store plus DB, that would also simplify things.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/SecureMultiStore.java
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/SecureMultiStore.java]: why is a salt needed?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/SecureMultiStore.java
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/SecureMultiStore.java]: do we need versions?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/SecureMultiStore.java
Comment (dli357) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/SecureMultiStore.java]: Removed versioning due to redesign.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/Util.java
Comment (dli357) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/Util.java]: Fixed, thanks!

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/CredentialIdentityStore.java
Comment (dli357) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/CredentialIdentityStore.java]: Made it package private, thanks!

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/SecureMultiStore.java
Comment (dli357) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/SecureMultiStore.java]: This existing design actually adds an index to make versioned writing possible to avoid data synchronization issues. However, after offline syncing, we decided to put all of the resources directly in a structured table instead of leveraging SecureStore.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/SecureMultiStore.java
Comment (dli357) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/SecureMultiStore.java]: This was originally needed because namespaces are case-sensitive, but secure keys are not. However, we've moved away from this design entirely, so this is no longer applicable.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/SecureMultiStore.java
Comment (dli357) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/SecureMultiStore.java]: Since we're moving away from using SecureStore, this is no longer needed.

File: cdap-proto/src/main/java/io/cdap/cdap/proto/credential/CredentialProfile.java
Comment (albertshau) [inline cdap-proto/src/main/java/io/cdap/cdap/proto/credential/CredentialProfile.java]: what would you think about changing the naming from credential provisioner to credential provider? It would match the current OAuth handler terminology more closely and also be more clearly different than cluster provisioners

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/CredentialIdentityStore.java
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/CredentialIdentityStore.java]: It's better to follow the abstractions we have in other places where it's http handler -> service -> table. The table layer just handles reads/writes, while the service layer starts transactions and performs operations across potentially multiple tables. This lets the credential table be used in a transaction with other tables.

See AppStateTable for an example.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/CredentialIdentityStore.java
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/CredentialIdentityStore.java]: it's more concise to use Optional's `map` method here. Similar comment throughout the class

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/CredentialIdentityStore.java
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/CredentialIdentityStore.java]: will we have separate create and update APIs? most of our REST APIs just have a PUT, which ends up being create or update in the same call.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/CredentialIdentityStore.java
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/CredentialIdentityStore.java]: nit: normally unmodifiable collections are used when returning a collection to prevent callers from modifying internal state. It shouldn't be needed if we're just creating keys.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/CredentialIdentityStore.java
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/CredentialIdentityStore.java]: it might be good to throw a more specific exception here, especially if we want this to eventually result in a 400 (I think IllegalArgument gets translated into a 500?)

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/CredentialProfileStore.java
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/CredentialProfileStore.java]: similar comment in this class about `map`

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/CredentialProfileStore.java
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/CredentialProfileStore.java]: similar comment in this class about having a Table layer that doesn't start a transaction itself

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/Utils.java
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/Utils.java]: nit: utility classes should have a more descriptive name like CredentialProfiles to try and keep them more focused.

File: cdap-proto/src/main/java/io/cdap/cdap/proto/credential/CredentialProfile.java
Comment (dli357) [reply cdap-proto/src/main/java/io/cdap/cdap/proto/credential/CredentialProfile.java]: Sounds reasonable, changed

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/CredentialIdentityStore.java
Comment (dli357) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/CredentialIdentityStore.java]: Done

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/CredentialIdentityStore.java
Comment (dli357) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/CredentialIdentityStore.java]: Done

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/CredentialProfileStore.java
Comment (dli357) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/CredentialProfileStore.java]: Done

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/CredentialIdentityStore.java
Comment (dli357) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/CredentialIdentityStore.java]: Done, refactored into a manager->table structure.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/CredentialIdentityStore.java
Comment (dli357) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/CredentialIdentityStore.java]: I'm keeping the APIs separate because it adds simplicity for the access control logic, otherwise we'll need to check different permissions depending on if the resource exists or not. It's possible to write the logic that way, but seems unnecessarily complex IMO. WDYT?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/CredentialIdentityStore.java
Comment (dli357) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/CredentialIdentityStore.java]: Done, changed this to BadRequestException

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/CredentialProfileStore.java
Comment (dli357) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/CredentialProfileStore.java]: Done

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/Utils.java
Comment (dli357) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/store/Utils.java]: Done

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/CredentialIdentityManager.java
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/CredentialIdentityManager.java]: this type of validation should happen higher up, whenever we have a http handler. Basically the handler should convert user provided strings into our internal java objects (CredentialIdentityId, etc) and make sure all user input is syntactically correct.  After that point, all the other layers should be able to assume names, ids, etc. are all valid.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/CredentialManagerUtils.java
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/CredentialManagerUtils.java]: this validation should happen as part of creating the CredentialIdentityId or CredentialProfileId instance so that it's impossible to create an invalid instance of an id class.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/CredentialIdentityManager.java
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/CredentialIdentityManager.java]: I think a 404 is actually more appropriate if the profile does not exist. 400 usually indicates there's something statically wrong with the request, like it is malformed, and that a retry would also result in a 400.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/CredentialIdentityManager.java
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/CredentialIdentityManager.java]: is this an interface because we'll have the default implementation and also a remote implementation?

If we won't have a remote one we can just have the concrete class and skip the interface.

**Review by albertshau — COMMENTED** [review]
please be sure to also fix the checkstyle warnings

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/CredentialManagerUtils.java
Comment (dli357) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/CredentialManagerUtils.java]: Fixed, I believe there was already duplicate logic there, made it a bit more strict. We'll catch IllegalArgumentException in the HTTP handler.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/CredentialIdentityManager.java
Comment (dli357) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/CredentialIdentityManager.java]: Fixed

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/CredentialIdentityManager.java
Comment (dli357) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/CredentialIdentityManager.java]: I think there may need to be a remote implementation for the namespace creation hook. I can remove the interface for the profile manager though.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/CredentialIdentityManager.java
Comment (dli357) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/CredentialIdentityManager.java]: Moved it into construction of the ID object, will catch IllegalArgumentException in the handler.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/CredentialIdentityManager.java
Comment (dli357) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/credential/CredentialIdentityManager.java]: On a second thought, I'll go ahead and remove the interfaces and add them back when necessary

**Review by albertshau — APPROVED** [review]
lgtm assuming all checkstyle warning are fixed before merging

**Pull request merged**
